expr/func_groupbytags_test.go

```diff
@@ -0,0 +1,349 @@
+}
+
+func TestGroupByTagsMultipleGroupByName(t *testing.T) {
+	in := []models.Series{
+		getModel("name1;tag1=val1;tag2=val2_0;tag3=3", a),
+		getModel("name1;tag1=val1;tag2=val2_1;tag3=3", b),
+		getModel("name2;tag1=val1_1;tag2=val2_0;tag3=3", c),
+		getModel("name2;tag1=val1_1;tag2=val2_1;tag3=3", d),
+	}
+	out := []models.Series{
+		getModel("name1", sumab),
+		getModel("name2", sumcd),
+	}
+
+	testGroupByTags("MultipleSeriesMultipleResultsMultipleNamesMoreTags", in, out, "sum", []string{"name"}, t)
+}
+
+func TestGroupByTagsMultipleSeriesMissingTag(t *testing.T) {
+	in := []models.Series{
+		getModel("name1;tag1=val1;tag2=val2_0", a),
+		getModel("name1;tag1=val1;tag2=val2_1", b),
+		getModel("name2;tag1=val1_1;tag2=val2_0", c),
+		getModel("name2;tag1=val1_1;tag2=val2_1", d),
+	}
+	out := []models.Series{
+		getModel("name1;missingTag=;tag1=val1", sumab),
+		getModel("name2;missingTag=;tag1=val1_1", sumcd),
+	}
+
+	testGroupByTags("MultipleSeriesMultipleResultsGroupByName", in, out, "sum", []string{"tag1", "name", "missingTag"}, t)
+}
+
+func TestGroupByTagsAllAggregators(t *testing.T) {
+	aggregators := []struct {
+		name             string
+		result1, result2 []schema.Point
+	}{
+		{name: "sum", result1: sumab, result2: sumabc},
+		{name: "avg", result1: avgab, result2: avgabc},
+		{name: "average", result1: avgab, result2: avgabc},
+		{name: "max", result1: maxab, result2: maxabc},
+		{name: "median", result1: medianab, result2: medianabc},
+		{name: "multiply", result1: multab, result2: multabc},
+		{name: "stddev", result1: stddevab, result2: stddevabc},
+		{name: "diff", result1: diffab, result2: diffabc},
+		{name: "range", result1: rangeab, result2: rangeabc},
+	}
+
+	for _, agg := range aggregators {
+		in := []models.Series{
+			getModel("name1;tag1=val1;tag2=val2_0", a),
+			getModel("name1;tag1=val1;tag2=val2_1", b),
+			getModel("name2;tag1=val1_1;tag2=val2_0", a),
+			getModel("name2;tag1=val1_1;tag2=val2_1", b),
+			getModel("name2;tag1=val1_1;tag2=val2_2", c),
+		}
+		out := []models.Series{
+			getModel("name1;tag1=val1", agg.result1),
+			getModel("name2;tag1=val1_1", agg.result2),
+		}
+
+		testGroupByTags("AllAggregators:"+agg.name, in, out, agg.name, []string{"tag1", "name"}, t)
+	}
+}
+
+func testGroupByTags(name string, in []models.Series, out []models.Series, agg string, tags []string, t *testing.T) {
+	f := NewGroupByTags()
+	gby := f.(*FuncGroupByTags)
+	gby.in = NewMock(in)
+	gby.aggregator = agg
+	gby.tags = tags
+
+	got, err := f.Exec(make(map[Req][]models.Series))
+	if err != nil {
+		t.Fatalf("case %q: err should be nil. got %q", name, err)
+	}
+	if len(got) != len(out) {
+		t.Fatalf("case %q: GroupByTags output expected to be %d but actually %d", name, len(out), len(got))
+	}
+
+	// Make sure got and out are in the same order
+	sort.Slice(got, func(i, j int) bool {
+		return got[i].Target < got[j].Target
+	})
+	sort.Slice(out, func(i, j int) bool {
+		return out[i].Target < out[j].Target
+	})
```

Comment:
is the order of `got` predictable ? (e.g sorted) ? does graphite sort it? if so, could we just pass in the correct `out` so that neither requires sorting?

Reply:
In my implementation, it's definitely not sorted (it's just in whatever order the map puts it). I'm not sure if graphite sorts it. I guess it would sort by target if anything? But if graphite does this, I imagine it would do it after functions complete. It wouldn't make sense to sort it in the middle of processing, when an `alias` function could come in and require re-sorting.

Reply:
right, makes sense